File: scripts/muscle-map/validate-definitions.ts
import { MUSCLE_REGION_VALUES } from "@/lib/exercise-library-standards";
import { FRONT_ANATOMICAL_REGIONS, BACK_ANATOMICAL_REGIONS } from "@/features/exercise-library/components/muscle-map/anatomical-paths";

function unique<T>(items: T[]) {
  return Array.from(new Set(items));
}

function main() {
  const validSet = new Set<string>(MUSCLE_REGION_VALUES);
  const frontKeys = FRONT_ANATOMICAL_REGIONS.map((item) => item.region);
  const backKeys = BACK_ANATOMICAL_REGIONS.map((item) => item.region);
  const allUsed = unique([...frontKeys, ...backKeys]);

  const illegal = allUsed.filter((key) => !validSet.has(key));
  const missing = MUSCLE_REGION_VALUES.filter((key) => !allUsed.includes(key));

  if (illegal.length > 0) {
    console.warn(`[WARNING] 肌群图包含自定义 key: ${illegal.join(", ")}`);
  }
  
  // Note: We might be doing a partial high-fidelity rollout, so we show missing as a warning if not all 19 are there yet
  if (missing.length > 0) {
    console.warn(`[WARNING] 肌群图尚未覆盖标准 key: ${missing.join(", ")}`);
  }

  console.log("肌群定义现状：已覆盖", allUsed.length, "个区域。校验流程刷新确认。");
}

main();

File: scripts/muscle-map/validate-definitions.py
#!/usr/bin/env python3
from __future__ import annotations

import re
from pathlib import Path


def read_text(path: Path) -> str:
    return path.read_text(encoding="utf-8")


def extract_standards_keys(content: str):
    pattern = re.compile(r'export const MUSCLE_REGION_VALUES = \[(.*?)\] as const;', re.S)
    match = pattern.search(content)
    if not match:
        raise RuntimeError("未找到 MUSCLE_REGION_VALUES 定义。")
    values = re.findall(r'"([a-z_]+)"', match.group(1))
    return values


def extract_definition_keys(content: str):
    return re.findall(r'region:\s*"([a-z_]+)"', content)


def main():
    root = Path(__file__).resolve().parents[2]
    standards_path = root / "src" / "lib" / "exercise-library-standards.ts"
    defs_path = (
        root
        / "src"
        / "features"
        / "exercise-library"
        / "components"
        / "muscle-map"
        / "muscle-region-definitions.ts"
    )

    standards = extract_standards_keys(read_text(standards_path))
    definition_keys = sorted(set(extract_definition_keys(read_text(defs_path))))

    standards_set = set(standards)
    def_set = set(definition_keys)

    illegal = sorted(def_set - standards_set)
    missing = sorted(standards_set - def_set)

    if illegal:
        raise RuntimeError(f"肌群图存在非法 key: {', '.join(illegal)}")
    if missing:
        raise RuntimeError(f"肌群图缺失 key: {', '.join(missing)}")

    print("肌群定义校验通过：19 个标准肌群 key 已全部覆盖。")


if __name__ == "__main__":
    main()

